lib/core.ml

```diff
@@ -178,7 +183,7 @@ let rec loop info ~colors ~positions s ~pos ~last st =
 
 let rec loop_no_mark info ~colors s ~pos ~last st =
   if pos < last then
-    let st' = st.next.(Char.code colors.[Char.code s.[pos]]) in
+    let st' = Array.unsafe_get st.next (Char.code colors.[Char.code s.[pos]]) in
```

Comment:
There are still some bound checks here.

Reply:
Ah indeed. Anyway, this PR is just for demonstration. I'm going to go ahead and rebase your original PR, since the changes apart from the bound checks do improve the performance on my laptop (though less than the bound checks).